nym-vpn/ui/src-tauri/src/commands/connection.rs

```diff
@@ -140,35 +239,57 @@ pub async fn disconnect(
 
     // switch to "Disconnecting" state
     app_state.state = ConnectionState::Disconnecting;
-    // unlock the mutex
-    drop(app_state);
+
     app.emit_all(
         EVENT_CONNECTION_STATE,
         ConnectionEventPayload::new(ConnectionState::Disconnecting, None, None),
     )
     .ok();
 
-    // TODO fake some delay to confirm disconnection
-    let app_state_cloned = state.inner().clone();
-    let task = tokio::spawn(async move {
-        sleep(Duration::from_secs(1)).await;
-        trace!("disconnected");
-
-        let mut state = app_state_cloned.lock().await;
-        state.state = ConnectionState::Disconnected;
-        state.connection_start_time = None;
+    let Some(ref mut vpn_tx) = app_state.vpn_ctrl_tx else {
+        app_state.state = ConnectionState::Disconnected;
+        app_state.connection_start_time = None;
+        app.emit_all(
+            EVENT_CONNECTION_STATE,
+            ConnectionEventPayload::new(
+                ConnectionState::Disconnected,
+                Some("vpn handle has not been initialized".to_string()),
+                None,
+            ),
+        )
+        .ok();
+        return Err(CmdError::new(
+            CmdErrorSource::InternalError,
+            "vpn handle has not been initialized".to_string(),
+        ));
+    };
 
-        debug!("sending event [{}]: disconnected", EVENT_CONNECTION_STATE);
+    // send Stop message to the VPN client
+    vpn_tx.send(NymVpnCtrlMessage::Stop).await.map_err(|e| {
+        let err_message = format!("failed to send Stop message to VPN client: {}", e);
+        error!(err_message);
         app.emit_all(
             EVENT_CONNECTION_STATE,
-            ConnectionEventPayload::new(ConnectionState::Disconnected, None, None),
+            ConnectionEventPayload::new(
+                ConnectionState::Disconnected,
+                Some(err_message.clone()),
+                None,
+            ),
         )
         .ok();
-    });
+        CmdError::new(CmdErrorSource::InternalError, err_message)
+    })?;
 
-    let _ = task.await;
+    app_state.state = ConnectionState::Disconnected;
```

Comment:
Set "disconnecting" here, no need to set disconnected, that is handled in the other task that listens for vpn stopping